docs/adr/adr-014-bootstrap-from-previous-peers.md

```diff
@@ -214,67 +210,67 @@ Such callbacks are easily passable as options at header module construction time
+With the ability to retrieve previously seen good peers, we should distinguish between those and the trusted peers for cases when we have expired local sync target (_i.e: subjective head_). On such cases, we would like to strictly sync and retrieve the subjective head from our trusted peers and no other, and for other syncing tasks, include the other stored good peers. To achieve this, we need to first include information about the expiry of the local sync target:
 
 ```diff
+
  // newP2PExchange constructs a new Exchange for headers.
- func newP2PExchange(cfg Config) func(
-        fx.Lifecycle,
-        ...
+ func newP2PExchange(
+        lc fx.Lifecycle,
++        storeChainHead *header.ExtendedHeader,
+        bpeers modp2p.Bootstrappers,
+        network modp2p.Network,
+        host host.Host,
         ...
-        []p2p.Option[p2p.ClientParameters],
-+       pstore peerstore.Peerstore,
- ) (libhead.Exchange[*header.ExtendedHeader], error) {
-        return func(
-                lc fx.Lifecycle,
-                ...
-                ...,
-                opts []p2p.Option[p2p.ClientParameters],
-        ) (libhead.Exchange[*header.ExtendedHeader], error) {
-                peers, err := cfg.trustedPeers(bpeers)
-                if err != nil {
-                        return nil, err
-                }
-+              list, err := pstore.Load()
-+               if err != nil {
-+                       return nil, err
-+               }
-+               peers := append(peers, list...)
-                ids := make([]peer.ID, len(peers))
-                for index, peer := range peers {
-                        ids[index] = peer.ID
-                        host.Peerstore().AddAddrs(peer.ID, peer.Addrs, peerstore.PermanentAddrTTL)
-                }
-                exchange, err := p2p.NewExchange[*header.ExtendedHeader](host, ids, conngater, opts...)
-                ...
-                return exchange, nil
-        }
- }
+        exchange, err := p2p.NewExchange(...,
+                p2p.WithParams(cfg.Client),
+                p2p.WithNetworkID[p2p.ClientParameters](network.String()),
+                p2p.WithChainID(network.String()),
++               p2p.WithSubjectiveInitialization(storeChainHead.IsExpired),
+        )
+        if err != nil {
+                return nil, err
 ```
 
-(_Code Snippet 4.a: Example of required changes to: celestia-node/nodebuilder/header/constructors.go_)
+(_Code Snippet 3.c: Example of required changes to celestia-node/nodebuilder/header/constructors.go_)
+
+This piece information will allow the `Exchange` to distinguish between the two cases and use the correct peer set for each case:
+
+```diff
+ func (ex *Exchange[H]) Head(ctx context.Context) (H, error) {
+        log.Debug("requesting head")
++       // pool of peers to request from is determined by whether the head of the store is expired or not
++       if ex.Params.subjectiveInitialisation {
```

Comment:
one thing we need to consider additionally that i forgot to mention on our call yesterday is that the Head method on exchange can get called during runtime (not just on start for setting the initial sync target). There should be a way to also determine during runtime whether Head should use all good peers or only trusted. It might have to become an argument for the method. Let's see if we can work around it as I'd prefer to not break interface.

Reply:
Yes indeed